### PR TITLE
Distributed systems evolution: message contracts

### PR DESCRIPTION
**Pull request by mauroservienti**
## PoA

- [x] Synopsis
- [x] Tags
- [x] Header image
- [x] Header image credits
- [x] Review
- [x] Rename file to match publish date
- [ ] Buffer content

### CODE REVIEW

File: _posts/messages-evolution.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+As I said, I work for [Particular Software](https://particular.net), the makers of [NServiceBus](https://docs.particular.net/tutorials/quickstart/), a messaging middleware aiming to ease the usage of messaging and queues when designing distributed systems.
+
+I'll assume the system we need to evolve is a message-based one. It's indeed true that many of the discussed techniques apply to other technologies like HTTP or gRPC, where the exchanged payload is assimilable to messages.
+
+## Evolving messages
+
+In the [previous article](https://milestone.topics.it/2022/06/11/distributed-systems-evolution-challenges.html), we discussed the direction of the changes. We mentioned that it's generally better to start applying changes from the inside and then gradually move to the outside.
+
+Applying the inside-to-outside technique to message evolution means changing receivers before publishers/senders. Let's take into account a subset of the scenario outlined above:
```
Comment (kbaley) [inline _posts/messages-evolution.md:23]:
```suggestion
Applying the inside-out technique to message evolution means changing receivers before publishers/senders. Let's take into account a subset of the scenario outlined above:
```

File: _posts/messages-evolution.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+<div class="mermaid">
+sequenceDiagram
+   Rest API ->> Backend: Queue message
+</div>
+
+We want to update the Backend endpoint first, and once that's in production, we can apply changes to the API endpoint.
+
+### Message exchange and contracts sharing
+
+At this point, we need a little digression and dive a bit into how messages are exchanged between endpoints. In the .NET world, there is a preference for strong typing. Developers prefer using C# classes, as presented above, to define messages. Those classes, which effectively are contracts, are generally shared across endpoints as NuGet packages. The `SubmitRequest` type is built into an assembly, packaged into a NuGet package, and referenced by the different endpoints using it to communicate.
```
Comment (kbaley) [inline _posts/messages-evolution.md:34]:
```suggestion
At this point, we need a little digression to dive a bit into how messages are exchanged between endpoints. In the .NET world, there is a preference for strong typing. Developers prefer using C# classes, as presented above, to define messages. Those classes, which effectively are contracts, are generally shared across endpoints as NuGet packages. The `SubmitRequest` type is built into an assembly, packaged into a NuGet package, and referenced by the different endpoints using it to communicate.
```

File: _posts/messages-evolution.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+### Where does the receivers' responsibility end?
+
+The last thing to address is how to handle situations like: we stopped supporting the message, but somehow, more are coming. We have done all that we described so far. However, when the time comes, and we remove the handler for `SubmitRequest` from the receiver endpoint, we notice that messages are still arriving by looking at logs. For example, if you are using NServiceBus, logs will contain an error message like the following:
+
+> No handlers could be found for message type: _message type_
+
+_Source: <https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Pipeline/Incoming/LoadHandlersConnector.cs#L27>_
+
+That means not all senders have been updated to use the new message version. Or there are still in-flight messages. Even though we removed the `SubmitRequest` from the NuGet package, there are no guarantees that other teams will update sender endpoints and stop using the old message type.
```
Comment (kbaley) [inline _posts/messages-evolution.md:100]:
```suggestion
That means not all senders have been updated to use the new message version. Or there are still in-flight messages. Even though we removed `SubmitRequest` from the NuGet package, there are no guarantees that other teams will update sender endpoints and stop using the old message type.
```

**Review by kbaley — COMMENTED** [review]
This was very well-done. Clearly explained and not many suggestions.

**Comment by mauroservienti** [comment]
> This was very well-done. Clearly explained and not many suggestions.

Thanks! Credit goes where credit is due: thank you!

**Pull request merged**